driver-core/src/main/java/com/datastax/driver/core/Metadata.java

```diff
@@ -94,27 +94,43 @@ synchronized void rebuildSchema(String keyspace, String table, ResultSet ks, Res
                 KeyspaceMetadata ksm = KeyspaceMetadata.build(ksRow);
 
                 if (cfDefs.containsKey(ksName)) {
-                    buildTableMetadata(ksm, cfDefs.get(ksName), colsDefs.get(ksName), cassandraVersion);
+                    buildTableMetadata(ksm, null, cfDefs.get(ksName), colsDefs.get(ksName), cassandraVersion);
                 }
                 addedKs.add(ksName);
-                keyspaces.put(ksName, ksm);
+
+                KeyspaceMetadata previousKsm = keyspaces.get(ksName);
+                if (previousKsm == null) {
+                    keyspaces.put(ksName, ksm);
+                    triggerOnKeyspaceAdded(ksm);
+                } else if (!previousKsm.equals(ksm)) {
+                    keyspaces.put(ksName, ksm);
+                    triggerOnKeyspaceChanged(ksm, previousKsm);
+                }
             }
 
-            // If keyspace is null, it means we're rebuilding from scratch, so
-            // remove anything that was not just added as it means it's a dropped keyspace
-            if (keyspace == null) {
-                Iterator<String> iter = keyspaces.keySet().iterator();
-                while (iter.hasNext()) {
-                    if (!addedKs.contains(iter.next()))
-                        iter.remove();
+            Set<String> removedKs = new HashSet<String>();
+            for (String ksName : keyspaces.keySet()) {
```

Comment:
Another killer argument in favor of not broadcasting events for tables in a dropped keyspace:  it seems that Cassandra already does exactly that :) : it first sends a DROPPED TABLE event for each table, then a DROPPED KEYSPACE. In the end, the user will get notifications for each table. That also explained why my tests were seeing duplicate events.